iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/CommandDetector.java

```diff
@@ -42,74 +52,155 @@ public static Builder builder() {
+        }
+        // Re-add predicate to stack to be reevaluated on the next argument
+        if (currentPredicate.is(ZERO_OR_MORE)) {
+          predicateStack.addFirst(currentPredicate);
+        }
+        // Add matched argument to command
+        commandArguments.add(resolution);
+      } else if (currentPredicate.is(OPTIONAL, ZERO_OR_MORE, NO_MATCH)) {
+        // Re-add argument to be evaluated by the next predicate
+        argumentStack.addFirst(resolution);
+      } else {
+        // Stop argument detection in case the argument does not match and the predicate is not optional or should not be matched
+        return Collections.emptyList();
       }
     }
-    return sizeCommand;
+    return commandArguments;
+  }
+
+  private static boolean remainingPredictsAreOptional(CommandPredicate currentPredicate, Deque<CommandPredicate> remainingPredicates) {
+    remainingPredicates.addFirst(currentPredicate);
+    return remainingPredicates.stream().noneMatch(predicate -> predicate.is(MATCH));
   }
 
   public static class Builder {
 
     List<CommandPredicate> predicates = new ArrayList<>();
 
-    private void addPredicate(Predicate<String> predicate, boolean optional) {
-      predicates.add(new CommandPredicate(predicate, optional));
+    private void addPredicate(Predicate<String> predicate, CommandPredicate.Type type) {
+      predicates.add(new CommandPredicate(predicate, type));
     }
 
     public CommandDetector.Builder with(Predicate<String> predicate) {
-      addPredicate(predicate, false);
+      addPredicate(predicate, MATCH);
       return this;
     }
 
+    public CommandDetector.Builder with(Collection<String> firstOf) {
+      return with(firstOf::contains);
+    }
+
+    public CommandDetector.Builder with(String expectedString) {
+      return with(expectedString::equals);
+    }
+
     public CommandDetector.Builder withOptional(Predicate<String> predicate) {
-      addPredicate(predicate, true);
+      addPredicate(predicate, OPTIONAL);
       return this;
     }
 
+    public CommandDetector.Builder notWith(Predicate<String> predicate) {
+      addPredicate(predicate, NO_MATCH);
+      return this;
+    }
+
+    public CommandDetector.Builder withOptionalRepeating(Predicate<String> predicate) {
+      addPredicate(predicate, ZERO_OR_MORE);
+      return this;
+    }
+
+    public CommandDetector.Builder withAnyFlagExcept(String... excludedFlags) {
+      List<String> excludedFlagList = Arrays.asList(excludedFlags);
+      return withOptionalRepeating(s -> s.startsWith("-") && !excludedFlagList.contains(s))
+        .notWith(excludedFlagList::contains);
+    }
+
+    public CommandDetector.Builder withAnyFlag() {
+      return withOptionalRepeating(s -> s.startsWith("-"));
+    }
+
     public CommandDetector build() {
       return new CommandDetector(predicates);
     }
+
+    public CommandDetector.Builder addPredicatesFromBuilder(CommandDetector.Builder otherBuilder) {
```

Comment:
Please move this method above the `build()` method so that the final method will always be the last. Additionally, maybe renaming the method to `withPredicatesFrom` would fit better in naming the other methods. WDYT?